tests/components/sonos/test_init.py

```diff
@@ -132,3 +140,254 @@ async def test_async_poll_manual_hosts_warnings(
+        ) as mock_async_call_later, patch(
+            "homeassistant.components.sonos.async_dispatcher_send"
+        ), patch(
+            "homeassistant.components.sonos.sync_get_visible_zones"
+        ) as mock_sync_get_visible_zones:
+            mock_sync_get_visible_zones.return_value = []
+            caplog.clear()
+
+            mock_discovery_message.side_effect = asyncio.TimeoutError("TimeoutError")
+            await manager.async_poll_manual_hosts()
+            assert len(caplog.messages) == 1
+            record = caplog.records[0]
+            assert record.levelname == "WARNING"
+            assert "Discovery message failed" in record.message
+            assert "TimeoutError" in record.message
+            mock_async_call_later.assert_called_once()
+
+
+async def test_async_poll_manual_hosts(hass: HomeAssistant) -> None:
+    """Tests the logic and execution branches for the function."""
+    await async_setup_component(
+        hass,
+        sonos.DOMAIN,
+        {"sonos": {"media_player": {"interface_addr": "127.0.0.1"}}},
+    )
+    await hass.async_block_till_done()
+    manager: SonosDiscoveryManager = hass.data[DATA_SONOS_DISCOVERY_MANAGER]
+
+    def _speaker_ping_success():
+        """Return success on speaker ping."""
+
+    def _speaker_ping_fail():
+        """Raise exception on speaker ping."""
+        raise SonosUpdateError()
+
+    with patch.object(manager, "_async_gethostbyname", side_effect=patch_gethostbyname):
+        manager.hosts.add("10.10.10.1")
+        manager.hosts.add("10.10.10.2")
+
+        # Test 1 first device fails, second device successful, speakers do not exist
+        with patch.object(
+            manager, "_async_handle_discovery_message", new=AsyncMock()
+        ) as mock_discovery_message, patch(
+            "homeassistant.components.sonos.async_call_later"
+        ) as mock_async_call_later, patch(
+            "homeassistant.components.sonos.async_dispatcher_send"
+        ), patch(
+            "homeassistant.components.sonos.sync_get_visible_zones"
+        ) as mock_sync_get_visible_zones:
+            mock_sync_get_visible_zones.side_effect = [OSError(), []]
+            await manager.async_poll_manual_hosts()
+            assert mock_sync_get_visible_zones.call_count == 2
+            assert mock_discovery_message.call_count == 1
+            assert mock_async_call_later.call_count == 1
+
+        # Test 2 first device successful, second device fails, speakers do not exist
+        with patch.object(
+            manager, "_async_handle_discovery_message", new=AsyncMock()
+        ) as mock_discovery_message, patch(
+            "homeassistant.components.sonos.async_call_later"
+        ) as mock_async_call_later, patch(
+            "homeassistant.components.sonos.async_dispatcher_send"
+        ), patch(
+            "homeassistant.components.sonos.sync_get_visible_zones"
+        ) as mock_sync_get_visible_zones:
```

Comment:
This is better but we should not patch the following:

https://developers.home-assistant.io/docs/development_testing?_highlight=writing#writing-tests-for-integrations

> Make sure to not interact with any integration details in tests of integrations. Following this pattern will make the tests more robust for changes in the integration.


`_async_handle_discovery_message`: this is integration code 
`async_call_later`: this is considered hass internals (events)
`async_dispatcher_send`: this is considered hass internals (events)
`sync_get_visible_zones`: this is integration code. Patch the pypi code instead (ex soco.household_id, soco.ui, soco.visible_zones)

Reply:
I'm on day 4 of hitting dead ends trying to do this and could use some help from someone with a better mastery of pytest fixtures than me. 

Main issue I'm having is current test fixture causes SoCo() to return the same singleton object with a fixed ip address - regardless of the IP address provided to the constructor.  For this test we need two independent SoCo objects with different IP addresses and the ability to mock their return values in the test code to hit the various scenarios.


https://github.com/home-assistant/core/blob/21d887dd0453d0cbbf3be42f4b3650ab92fc038d/tests/components/sonos/conftest.py#L87

Reply:
You can patch the patch target to be an iterable

If side_effect is an iterable then each call to the mock will return the next value from the iterable.

https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.side_effect

Reply:
Thanks, i ended up patching in a factory w/ cache so that when a soco object is being created 
```
soco = SoCo(ip_addr)
```
It'll return one whose ip address is what is specified in the call.  In test setup these can get pushed into the factory cache so the side effects for an ip_address are clear. For example if 10.10.10.1 is offline and 10.10.10.2 is online, then a speaker should get generated for 10.10.10.2 and an error message should be logged for 10.10.10.1.

Reply:
@bdraco - Success! The unit tests are definitely alot better and more concise using your recommended technique of mocking Soco.  Thanks for the feedback.  Please review when you have time.